File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode buildTree(int[] pre, int[] in) {
        Map<Integer, Integer> map = new HashMap<>();
        for(int i=0; i<in.length; i++) map.put(in[i], i);
        return helper(pre, new int[1], 0, pre.length-1, map);
    }

    private TreeNode helper(int[] pre, int[] index, int left, int right, Map<Integer, Integer> map) {
        if (left>right) return null;
        int rootValue = pre[index[0]++];
        TreeNode root = new TreeNode(rootValue);
        root.left = helper(pre, index, left, map.get(rootValue)-1, map);
        root.right = helper(pre, index, map.get(rootValue)+1, right, map);
        return root;
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameterOfBinaryTree(root) {
        if(!root) return 0;

        let max = 0;

        const getDiameter = (root) => {
            if(!root) return 0;

            let left = getHeight(root.left);
            let right = getHeight(root.right);
            console.log(root.val, left, right, max);
            // if(!root.left) {
            //     max = Math.max(max, right + 1);
            //     return max;
            // }

            // if(!root.right){
            //     max = Math.max(max, left + 1);
            //     return max;
            // }
            // console.log('before both',root.val, left, right, max);
            
            // if(root.left && root.right)
                max = Math.max(max, left + right, getDiameter(root.left), getDiameter(root.right));
            return max;
        }

        const getHeight = (root) => {
            if(!root) return 0;

            return Math.max(getHeight(root.left), getHeight(root.right)) + 1;
        }

        return getDiameter(root);

        return max;
    }
}
